### PR TITLE
Integrate api with login page

### PR DESCRIPTION
**Pull request by danielduong8**
Refactor state hooks and integrate api 

- Refactored state hooks for cleaner code

- Add form validation before calling api

- Install axios

Setup navigation on successful login

Add api class

Add MaterialUI Alert components

### CODE REVIEW

File: client/src/api/v1/index.tsx
Comment (andy9775) [inline client/src/api/v1/index.tsx:37]: You don't want to navigate here. This should only be network related stuff.

File: client/src/containers/AuthContainer.test.tsx
Comment (andy9775) [inline client/src/containers/AuthContainer.test.tsx:93]: Better test name

File: client/src/containers/AuthContainer.test.tsx
Comment (andy9775) [inline client/src/containers/AuthContainer.test.tsx:156]: This is perfect. What you could also do, 1/ is assert that it was called with the expected parameters. 2/ return some mock data then assert that is is displayed in the UI as expected (this is more relevant for the `posts` logic, but I'm leaving it here for reference).

File: client/src/containers/AuthContainer.test.tsx
Comment (andy9775) [inline client/src/containers/AuthContainer.test.tsx:164]: If you assert that mockSignup was called with certain params (the expected ones) you can avoid mockAlert and asserting that it was triggered. Testing the call params is more robust.

File: client/src/containers/AuthContainer.test.tsx
Comment (andy9775) [inline client/src/containers/AuthContainer.test.tsx:169]: Successful login test?

File: client/src/containers/AuthContainer.test.tsx
Comment (andy9775) [inline client/src/containers/AuthContainer.test.tsx:116]: You can assert that the mockApi function was not called rather than asserting that the alert was called with some error. If the error message changes for example, that would cause the test to fail when it really shouldn't (since there error message is just a visual thing, whereas this is a logic test).

**Review by andy9775 — CHANGES_REQUESTED** [review]
I don't think the api client tests are really necessary. I would also avoid `window.error`. 
Otherwise good job, some some minor fixes/comments.

File: client/src/containers/AuthContainer.test.tsx
Comment (danielduong8) [reply client/src/containers/AuthContainer.test.tsx:164]: How would we test the different params since we can't actually test what happens to mockSignUp based on these different inputs?

Also, I'm not sure how to differentiate a successful sign up from a failed signup without checking the alert messages logged.

Comment (andy9775) [reply client/src/containers/AuthContainer.test.tsx:164]: `expect(mockSignUp).toBeCalledWith(expectedEmail, expectedUserName, ...)` where expectedEmail is the email that the form input was set to. So if in your test, you set the users email in the form to be 'foo@mail.utoronto.ca`, the mock signup function should be called with  'foo@mail.utoronto.ca` as the first param (similar logic extends to username, password, passwordConfirmation, first and last name). If the mock signup function has not been called, that means that some component level validation prevented that from happening (e.g. the email format was bad). 

So if the mock signup function was called, that means that it was a successful signup. If there was not successful signup, e.g. email format was invalid, or the password was too short, then the mockSignup function should not be called.

**Comment by danielduong8** [comment]
The last commit is just renaming a bunch of "login" to "signin"; you can just skip that commit and look at the previous one to see the actual changes.

I'm sorry I just had to lol, it was bothering me greatly.

File: client/src/components/SignIn.tsx
Comment (andy9775) [inline client/src/components/SignIn.tsx:38]: isFormMissingFields

File: client/src/components/SignUp.tsx
Comment (andy9775) [inline client/src/components/SignUp.tsx:33]: isFormMissingFields

File: client/src/containers/AuthContainer.test.tsx
Comment (andy9775) [inline client/src/containers/AuthContainer.test.tsx:165]: Perfect

File: client/src/containers/AuthContainer.test.tsx
Comment (andy9775) [inline client/src/containers/AuthContainer.test.tsx:173]: I would rename this to be something more descriptive of the process being tested, not some inner function name while we're here.

File: client/src/containers/AuthContainer.test.tsx
Comment (andy9775) [inline client/src/containers/AuthContainer.test.tsx:187]: I would add a test for a backend response indicating a bad password or something

File: client/src/api/v1/index.ts
Comment (andy9775) [inline client/src/api/v1/index.ts:10]: I would make the return type a generic as well. E.g. sign-in IIRC returns some user info.

https://www.oreilly.com/library/view/learning-typescript-2x/9781788391474/b170884f-368a-417e-9ba4-b5e86207b8dd.xhtml

**Review by andy9775 — APPROVED** [review]
I noticed that your taking an alert function as a prop, you coud combine it with https://reactjs.org/docs/portals.html you could just have a set of functions you import: `alertSuccess(msg)`, `alertError(msg)` and so on.  Alternatively, you could use https://reactjs.org/docs/context.html, have the context provider and alert component at the root (App.tsx) and then each  sub-component can just useContext (https://reactjs.org/docs/context.html) and you could have a bunch of alert functions passed in that way. This would be much cleaner and more reusable across different components, but I'll leave it up to you if you want to go that route. 

Just a few comments, otherwise LGTM! Feel free to submit once you address them

**Review by plausibly — APPROVED** [review]

**Review by plausibly — CHANGES_REQUESTED** [review]
Could you setup the "Logout" button on the post dashboard?

Daniel: Yes but probably in a separate PR.

**Pull request merged**
